src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+use rustc::mir::repr::{self, Mir};
+
+use std::fmt::Debug;
+use std::io;
+use std::marker::PhantomData;
+use std::mem;
+use std::path::PathBuf;
+use std::usize;
+
+use super::MirBorrowckCtxtPreDataflow;
+use super::gather_moves::{Location, MoveData, MovePathData, MovePathIndex, MoveOutIndex, PathMap};
+use super::gather_moves::{MoveOut, MovePath};
+use super::DropFlagState;
+
+use bitslice::BitSlice; // adds set_bit/get_bit to &[usize] bitvector rep.
+
+pub use self::sanity_check::sanity_check_via_rustc_peek;
+
+mod graphviz;
+mod sanity_check;
+
+pub trait Dataflow {
+    fn dataflow(&mut self);
+}
+
+impl<'a, 'tcx: 'a, BD> Dataflow for MirBorrowckCtxtPreDataflow<'a, 'tcx, BD>
+    where BD: BitDenotation, BD::Bit: Debug, BD::Ctxt: HasMoveData<'tcx>
+{
+    fn dataflow(&mut self) {
+        self.flow_state.build_sets();
+        self.pre_dataflow_instrumentation().unwrap();
+        self.flow_state.propagate();
+        self.post_dataflow_instrumentation().unwrap();
+    }
+}
+
+struct PropagationContext<'b, 'a: 'b, 'tcx: 'a, O>
+    where O: 'b + BitDenotation, O::Ctxt: HasMoveData<'tcx>,
+{
+    builder: &'b mut DataflowAnalysis<'a, 'tcx, O>,
+    changed: bool,
+}
+
+impl<'a, 'tcx: 'a, BD> DataflowAnalysis<'a, 'tcx, BD>
+    where BD: BitDenotation, BD::Ctxt: HasMoveData<'tcx>
+{
+    fn propagate(&mut self) {
+        let mut temp = vec![0; self.flow_state.sets.words_per_block];
+        let mut propcx = PropagationContext {
+            builder: self,
+            changed: true,
+        };
+        while propcx.changed {
+            propcx.changed = false;
+            propcx.reset(&mut temp);
+            propcx.walk_cfg(&mut temp);
+        }
+    }
+
+    fn build_sets(&mut self) {
+        // First we need to build the entry-, gen- and kill-sets. The
+        // gather_moves information provides a high-level mapping from
+        // mir-locations to the MoveOuts (and those correspond
+        // directly to gen-sets here). But we still need to figure out
+        // the kill-sets.
+
+        {
+            let sets = &mut self.flow_state.sets.for_block(repr::START_BLOCK.index());
+            self.flow_state.operator.start_block_effect(&self.ctxt, sets);
+        }
+
+        for bb in self.mir.all_basic_blocks() {
+            let &repr::BasicBlockData { ref statements,
+                                        ref terminator,
+                                        is_cleanup: _ } =
+                self.mir.basic_block_data(bb);
+
+            let sets = &mut self.flow_state.sets.for_block(bb.index());
+            for j_stmt in 0..statements.len() {
+                self.flow_state.operator.statement_effect(&self.ctxt, sets, bb, j_stmt);
+            }
+
+            if terminator.is_some() {
+                let stmts_len = statements.len();
+                self.flow_state.operator.terminator_effect(&self.ctxt, sets, bb, stmts_len);
+            }
+        }
+    }
+}
+
+fn on_all_children_bits<Each>(path_map: &PathMap,
```

Comment:
I think we should kill this and use `rustc_borrock::borrowck::mir::on_all_children_bits` instead, or at least comment on the discrepancy if there is one. I didn't touch it because I was not interested in figuring out `MovingOutStatements`.